Encoder/DefaultEncoder.php

```diff
@@ -0,0 +1,163 @@
+     * @var string
+     */
+    protected $encryptionEngine;
+
+    /**
+     * @var string
+     */
+    protected $encryptionAlgorithm;
+
+    /**
+     * Constructor.
+     *
+     * @param KeyLoaderInterface $keyLoader
+     * @param string             $encryptionEngine
+     * @param string             $encryptionAlgorithm
+     *
+     * @throws InvalidArgumentException If the given encryptionEngine is not supported.
+     * @throws InvalidArgumentException If the given algorithm is not supported.
+     */
+    public function __construct(KeyLoaderInterface $keyLoader, $encryptionEngine, $encryptionAlgorithm)
+    {
+        if (!in_array($encryptionEngine, ['OpenSSL', 'SecLib'])) {
+            throw new InvalidArgumentException(
+                sprintf('The encryption engine "%s" is not supported', $encryptionEngine)
+            );
+        }
+
+        if (!$this->isAlgorithmSupportedForEngine($encryptionEngine, $encryptionAlgorithm)) {
+            throw new InvalidArgumentException(
+                sprintf('The algorithm "%s" is not supported for %s', $encryptionAlgorithm, $encryptionEngine)
+            );
+        }
+
+        $this->keyLoader           = $keyLoader;
+        $this->encryptionEngine    = $encryptionEngine;
+        $this->encryptionAlgorithm = $encryptionAlgorithm;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \RuntimeException If the JWS cannot be signed
+     */
+    public function encode(array $data)
+    {
+        $jws = new JWS(['alg' => $this->encryptionAlgorithm], $this->encryptionEngine);
+        $key = $this->keyLoader->loadKey('private');
+
+        $jws->setPayload($data);
+
+        try {
+            $jws->sign($key, $this->keyLoader->getPassphrase());
+        } catch (InvalidArgumentException $e) {
+            throw new \RuntimeException(
+                sprintf('An error occurred while signing the JWS (%s). Please check your configuration.', $e->getMessage())
+            );
+        }
+
+        return $jws->getTokenString();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws InvalidArgumentException If the JWS cannot be loaded
+     * @throws \RuntimeException        If the JWS cannot be verified
+     */
+    public function decode($token)
+    {
+        try {
+            $jws = JWS::load($token, false, null, $this->encryptionEngine);
+        } catch (InvalidArgumentException $e) {
+            return false;
+        }
+
+        $key     = $this->keyLoader->loadKey('public');
+        $payload = $jws->getPayload();
+
+        try {
+            if (!$jws->verify($key, $this->encryptionAlgorithm) || $this->isJWSExpired($payload)) {
+                return false;
+            }
+        } catch (InvalidArgumentException $e) {
+            throw new \RuntimeException(
+                sprintf('An error occurred while verifying the JWS (%s). Please check your configuration.', $e->getMessage())
+            );
+        }
+
+        return $payload;
+    }
```

Comment:
Hmm, so if we make this change (which I think we should), we're actually slightly changing the interface of both `JWTEncoderInterface::decode` and `JWTManagerInterface::decode()` - as these *appear* to not advertise that an exception will ever be thrown. Because you kept BC so well with the original `JWTEncoder`, you were able to not update `JTWProvider` at all, which is pretty cool :).

So we need to be pragmatic to not make too many breaks... but make them when it has benefit. Technically, in the Symfony core, we'd probably make 2 new interfaces and deprecate the old ones... but that's overkill. I think we should:

A) Update the phpdoc of the interfaces to say that a certain type(s) of exception will be thrown. Make this as a BC break - perhaps add a new CHANGELOG.md file to the repo (a good idea anyways, so we can document the upgrade path).
B) Update (in this PR) this class to throw the exceptions
C) Then obviously, update JWTProvider

We could do this in 2 PR's, but it seems quite related, so let's get it done on one spot I think.